src/ML-PACE/pair_pace_extrapolation.cpp

```diff
@@ -0,0 +1,453 @@
+
+  // downwards modified by YL
+
+  double **x = atom->x;
+  double **f = atom->f;
+  tagint *tag = atom->tag;
+  int *type = atom->type;
+  // number of atoms in cell
+  int nlocal = atom->nlocal;
+
+  int newton_pair = force->newton_pair;
+
+  // number of atoms including ghost atoms
+  int nall = nlocal + atom->nghost;
+
+  // inum: length of the neighborlists list
+  inum = list->inum;
+
+  // ilist: list of "i" atoms for which neighbor lists exist
+  ilist = list->ilist;
+
+  //numneigh: the length of each these neigbor list
+  numneigh = list->numneigh;
+
+  // the pointer to the list of neighbors of "i"
+  firstneigh = list->firstneigh;
+
+  // this happens when used as substyle in pair style hybrid.
+  // So this check and error effectively disallows use with pair style hybrid.
+  if (inum != nlocal) { error->all(FLERR, "inum: {} nlocal: {} are different", inum, nlocal); }
+
+  //if flag_compute_extrapolation_grade at this iteration then
+  // grow extrapolation_grade_gamma array, that store per-atom extrapolation grades
+  if (flag_compute_extrapolation_grade && atom->nlocal > nmax) {
+    memory->destroy(extrapolation_grade_gamma);
+    nmax = atom->nlocal;
+    memory->create(extrapolation_grade_gamma, nmax, "pace/atom:gamma");
+    //zeroify array
+    memset(extrapolation_grade_gamma, 0, nmax * sizeof(*extrapolation_grade_gamma));
+  }
+
+  //determine the maximum number of neighbours
+  int max_jnum = 0;
+  int nei = 0;
+  for (ii = 0; ii < list->inum; ii++) {
+    i = ilist[ii];
+    jnum = numneigh[i];
+    nei = nei + jnum;
+    if (jnum > max_jnum) max_jnum = jnum;
+  }
+
+  if (flag_compute_extrapolation_grade)
+    aceimpl->ace->resize_neighbours_cache(max_jnum);
+  else
+    aceimpl->rec_ace->resize_neighbours_cache(max_jnum);
+
+  //loop over atoms
+  for (ii = 0; ii < list->inum; ii++) {
+    i = list->ilist[ii];
+    const int itype = type[i];
+
+    const double xtmp = x[i][0];
+    const double ytmp = x[i][1];
+    const double ztmp = x[i][2];
+
+    jlist = firstneigh[i];
+    jnum = numneigh[i];
+
+    // checking if neighbours are actually within cutoff range is done inside compute_atom
+    // mapping from LAMMPS atom types ('type' array) to ACE species is done inside compute_atom
+    //      by using 'ace->element_type_mapping' array
+    // x: [r0 ,r1, r2, ..., r100]
+    // i = 0 ,1
+    // jnum(0) = 50
+    // jlist(neigh ind of 0-atom) = [1,2,10,7,99,25, .. 50 element in total]
+    try {
+      if (flag_compute_extrapolation_grade)
+        aceimpl->ace->compute_atom(i, x, type, jnum, jlist);
+      else
+        aceimpl->rec_ace->compute_atom(i, x, type, jnum, jlist);
+    } catch (std::exception &e) {
+      error->one(FLERR, e.what());
+    }
+    // 'compute_atom' will update the `ace->e_atom` and `ace->neighbours_forces(jj, alpha)` arrays and max_gamma_grade
+
+    if (flag_compute_extrapolation_grade) {
+      //FIXME: need an updated PACE library URL since this currently fails compiling
```

Comment:
I have commented this line out to be able to compile the code with the referenced external library.
Obviously that one needs to be updated so that this commented line can be reactivated.

Reply:
after switching to use the CMakeLists.txt file bundled with the downloaded pace library, this hack is no longer needed.